sae/vm_test.go

```diff
@@ -270,3 +274,94 @@ func TestSyntacticBlockChecks(t *testing.T) {
 		})
 	}
 }
+
+func TestSemanticBlockChecks(t *testing.T) {
+	const now = 1e6
+	ctx, sut := newSUT(t, 1, blockstest.WithTimestamp(now))
+	sut.rawVM.config.Now = func() time.Time {
+		return time.Unix(now, 0)
+	}
+
+	lastAcceptedID, err := sut.LastAccepted(t.Context())
+	require.NoError(t, err, "LastAccepted()")
+	lastAccepted, err := sut.GetBlock(t.Context(), lastAcceptedID)
+	require.NoError(t, err, "GetBlock(lastAcceptedID)")
+
+	tests := []struct {
+		name           string
+		parentHash     common.Hash
+		acceptedHeight bool
+		time           uint64
+		receipts       types.Receipts
+		wantErr        error
+	}{
+		{
+			name:       "unknown_parent",
+			parentHash: common.Hash{1},
+			wantErr:    errUnknownParent,
+		},
+		{
+			name:           "already_finalized",
+			acceptedHeight: true,
+			wantErr:        errFinalizedBlock,
+		},
+		{
+			name:    "block_time_under_minimum",
+			time:    1,
+			wantErr: errBlockTimeUnderMinimum,
+		},
+		{
+			name:    "block_time_before_parent",
+			time:    sut.genesis.BuildTime() - 1,
+			wantErr: errBlockTimeBeforeParent,
+		},
+		{
+			name:    "block_time_after_maximum",
+			time:    now + uint64(maxFutureBlockTime.Seconds()) + 1,
+			wantErr: errBlockTimeAfterMaximum,
+		},
+		{
+			name: "hash_mismatch",
+			receipts: types.Receipts{
+				&types.Receipt{}, // Unexpected receipt
+			},
+			wantErr: errHashMismatch,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.parentHash == (common.Hash{}) {
+				tt.parentHash = common.Hash(lastAcceptedID)
+			}
+			height := lastAccepted.Height()
+			if !tt.acceptedHeight {
+				height++
+			}
+			if tt.time == 0 {
+				tt.time = now
+			}
+
+			ethB := types.NewBlock(
+				&types.Header{
+					ParentHash: tt.parentHash,
+					Number:     new(big.Int).SetUint64(height),
+					Time:       tt.time,
+				},
+				nil,
+				nil,
+				tt.receipts,
+				saetest.TrieHasher(),
+			)
+			b := blockstest.NewBlock(
+				t,
+				ethB,
+				nil,
+				nil,
+			)
+			snowB, err := sut.ParseBlock(ctx, b.Bytes())
+			require.NoErrorf(t, err, "ParseBlock(...)")
+			require.ErrorIs(t, snowB.Verify(t.Context()), tt.wantErr, "Verify()")
```

Comment:
The alternative was:
```suggestion
			err := sut.rawVM.VerifyBlock(t.Context(), nil, b)
			require.ErrorIs(t, err, tt.wantErr, "VerifyBlock()")
```
But I wasn't sure if that was preferred since it wasn't exposed as a method on `sut`.

Reply:
I deliberately do everything via the methods that consensus will see so that we get a full integration test, including of the adaptor.